src/main/java/io/swisschain/services/RootKeyRotationService.java

```
@@ -53,88 +58,174 @@ public void rotate(SymmetricKey currentKey, SymmetricKey previousKey) throws Exc
+        }
+        if (wallet.getRootKeyHash().equals(currentKeyHash)
+            || wallet.getRootKeyHash().equals(previousKeyHash)) {
+          continue;
         }
-        wallets = walletRepository.get(wallets.get(wallets.size() - 1).getId(), batchSize);
+
+        logger.error(
+            String.format(
+                "Wallet %d has unknown root key hash %s", wallet.getId(), wallet.getRootKeyHash()));
+
+        return false;
       }
 
-      rootKey = RootKey.create(currentKeyHash);
+      count += wallets.size();
 
-      rootKeyRepository.insert(rootKey);
+      logger.info(String.format("Processed %d", count));
 
-      logger.info("Root key rotation finished");
-    } else {
-      logger.info("Root key rotation is not needed");
+      wallets = walletRepository.get(wallets.get(wallets.size() - 1).getId(), batchSize);
     }
+
+    logger.info("Root keys validation finished");
+
+    return true;
+  }
+
+  private boolean rotate(
+      String currentKeyHash,
+      byte[] currentSecret,
+      byte[] currentNonce,
+      String previousKeyHash,
+      byte[] previousSecret,
+      byte[] previousNonce)
+      throws OperationFailedException, SQLException, OperationExhaustedException {
+    diagnosticService.report("Root key rotation in progress");
+    logger.info("Root key rotation started");
+
+    var count = 0;
+    var wallets = walletRepository.get(0, batchSize);
+
+    while (wallets.size() > 0) {
+      for (var wallet : wallets) {
+        if (currentKeyHash.equals(wallet.getRootKeyHash())) {
+          continue;
+        }
+
+        byte[] privateKey;
+
+        if (previousSecret != null && previousNonce != null) {
+          if (!previousKeyHash.equals(wallet.getRootKeyHash())) {
+            logger.error(
+                String.format(
+                    "Wallet %d has unknown root key hash %s. Root key rotation aborted.",
+                    wallet.getId(), wallet.getRootKeyHash()));
+            return false;
```

It would be more straightforward to throw an exception with a detailed description here instead of returning `false` and processing it on the topper level.